Vanguard Project - Vanguard Digital Experience Analysis

Dataset - https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_2.txt

In [1]:
import pandas as pd

url_pt2 = "https://raw.githubusercontent.com/data-bootcamp-v4/lessons/refs/heads/main/5_6_eda_inf_stats_tableau/project/files_for_project/df_final_web_data_pt_2.txt"

df2 = pd.read_csv(url_pt2, sep=",")

df2.head(3)

,client_id,visitor_id,visit_id,process_step,date_time
0,763412,601952081_10457207388,397475557_40440946728_419634,confirm,2017-06-06 08:56:00
1,6019349,442094451_91531546617,154620534_35331068705_522317,confirm,2017-06-01 11:59:27
2,6019349,442094451_91531546617,154620534_35331068705_522317,step_3,2017-06-01 11:58:48


In [2]:
# Size of dataset

print(df2.shape)

(412264, 5)


In [3]:
# Check column names and data types

df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 412264 entries, 0 to 412263
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype
---  ------        --------------   -----
 0   client_id     412264 non-null  int64
 1   visitor_id    412264 non-null  str  
 2   visit_id      412264 non-null  str  
 3   process_step  412264 non-null  str  
 4   date_time     412264 non-null  str  
dtypes: int64(1), str(4)
memory usage: 15.7 MB


In [4]:
# Check for missing values

df2.isnull().sum()

client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

In [5]:
# Check for duplicate rows

print(df2.duplicated().sum()) 

8669


In [6]:
# Actual duplicate rows 

df2[df2.duplicated()] 

,client_id,visitor_id,visit_id,process_step,date_time
125,3751049,937925261_74821699083,628817270_45049432060_475841,start,2017-05-15 23:09:44
143,9912156,746623513_52101087617,780044921_57678038107_984942,start,2017-05-03 14:09:52
326,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:49:09
328,5786602,893139880_61416949586,747153834_69434320996_858907,start,2017-05-10 01:48:45
375,5786602,893139880_61416949586,119968876_5423296886_390511,start,2017-06-08 02:14:09
...,...,...,...,...,...
412120,1672289,658797407_91003097342,430110748_90347503221_597628,start,2017-05-01 19:28:53
412163,4408245,14058946_73585431797,994148957_84373994761_991865,start,2017-06-06 09:17:23
412165,4408245,14058946_73585431797,994148957_84373994761_991865,start,2017-06-06 09:09:44
412167,4408245,14058946_73585431797,994148957_84373994761_991865,start,2017-06-06 08:51:16


In [7]:
# Check unique values
df2['process_step'].unique() 

<StringArray>
['confirm', 'step_3', 'step_2', 'step_1', 'start']
Length: 5, dtype: str

In [9]:
# Rows per stage ?
df2['process_step'].value_counts()  

process_step
start      135035
step_1      89761
step_2      71294
step_3      58614
confirm     57560
Name: count, dtype: int64

In [10]:
# just checking column date

df2['date_time'].dtype

<StringDtype(storage='python', na_value=nan)>

In [12]:
# just checking column date

df2['date_time'].dtype

print("Missing:", df2['date_time'].isna().sum())

print("Unique:", df2['date_time'].nunique())

print("Duplicates:", df2['date_time'].duplicated().sum())

Missing: 0
Unique: 346697
Duplicates: 65567


Which identifier should be used for each type of analysis?

In [13]:
# Check for unique values for column client, visitor and visit

print("Unique client_id:", df2["client_id"].nunique())
print("Unique visitor_id:", df2["visitor_id"].nunique())
print("Unique visit_id:", df2["visit_id"].nunique())
print("Rows:", len(df2))

Unique client_id: 67430
Unique visitor_id: 71042
Unique visit_id: 82841
Rows: 412264


One client has many visit --> Unique client_id: 67,430 != Unique visit_id: 82,841

In [14]:
# Check if one client has multiple visits

df2.groupby("client_id")["visit_id"].nunique().sort_values(ascending=False).head()

client_id
244312     17
1771611    15
8167818    13
561448     10
1198324    10
Name: visit_id, dtype: int64

Customer journey --> A client can return multiple times and create multiple visits.

In [25]:
# Client --> Visitor

df2.groupby("client_id")["visitor_id"].nunique().value_counts().sort_index()

visitor_id
1     63542
2      3467
3       335
4        63
5        12
6         6
7         2
9         1
11        1
13        1
Name: count, dtype: int64

Verifying Hierarchy 

In [27]:
# Visit --> Client 

df2.groupby("visit_id")["client_id"].nunique().value_counts()

client_id
1    82219
2      617
3        5
Name: count, dtype: int64

A visit should ideally belong to only one client.

In [28]:
# Visit --> Visitor 

df2.groupby("visit_id")["visitor_id"].nunique().value_counts()

visitor_id
1    82841
Name: count, dtype: int64

One visit id per visitor

In [26]:
# one visitor belongs to multiple clients

df2.groupby("visitor_id")["client_id"].nunique().value_counts()

client_id
1    70216
2      819
3        6
4        1
Name: count, dtype: int64

In [ ]:
Visitor_id are associated with multiple clients therefore should be avoided as a primary customer identifier (could be due to shared devices, authentication, household accounts) 

In [16]:
# Checking for exact duplicates

df2[df2.duplicated(keep=False)].sort_values(
["client_id", "visit_id"]
)

,client_id,visitor_id,visit_id,process_step,date_time
288212,1531,934069404_34543643308,110350144_67985219545_904677,start,2017-06-02 22:32:28
288213,1531,934069404_34543643308,110350144_67985219545_904677,start,2017-06-02 22:32:28
288214,1531,934069404_34543643308,110350144_67985219545_904677,start,2017-06-02 22:32:08
288215,1531,934069404_34543643308,110350144_67985219545_904677,start,2017-06-02 22:32:08
296377,2078,585735301_21309149782,900598259_99565669243_552843,start,2017-05-06 08:37:14
...,...,...,...,...,...
295097,9995203,550384864_76045959210,823630151_22344597498_646045,start,2017-06-11 12:23:43
85991,9996877,83108209_22162646095,713500557_31297619237_860300,start,2017-05-03 18:08:32
85992,9996877,83108209_22162646095,713500557_31297619237_860300,start,2017-05-03 18:08:32
63673,9998447,747535871_38029188908,309540451_47388672022_606106,start,2017-06-03 19:17:15


In [17]:
# Check repeated process steps within a visit

duplicates_steps = df2[
    df2.duplicated(
        subset=["client_id", "visit_id", "process_step"],
        keep=False
)
]

duplicates_steps.sort_values(
["client_id", "visit_id", "process_step", "date_time"]
)

,client_id,visitor_id,visit_id,process_step,date_time
285315,336,64757908_3400128256,649044751_80905125055_554468,start,2017-06-01 07:26:55
285190,336,64757908_3400128256,649044751_80905125055_554468,start,2017-06-01 07:42:43
100903,786,439747392_30293505155,640294664_79662229470_499394,start,2017-06-01 11:00:22
100902,786,439747392_30293505155,640294664_79662229470_499394,start,2017-06-01 11:01:20
251290,805,831412807_82548325803,451173196_11661340552_563345,start,2017-06-15 19:09:13
...,...,...,...,...,...
60662,9999768,389637884_26725643037,85676722_11636430786_122704,step_2,2017-06-03 18:11:39
60668,9999768,389637884_26725643037,85676722_11636430786_122704,step_3,2017-06-03 18:07:24
60661,9999768,389637884_26725643037,85676722_11636430786_122704,step_3,2017-06-03 18:12:43
305392,9999875,738878760_1556639849,931268933_219402947_599432,step_1,2017-06-01 22:40:08


In [22]:
# How often it happens
(
df2.groupby(["client_id", "visit_id", "process_step"])
    .size()
    .sort_values(ascending=False)
    .head(20)
)

client_id  visit_id                      process_step
913770     512831152_67980938484_874337  start           48
5574879    875138661_34710212496_881092  step_1          40
8312319    356549486_83727228650_639957  start           33
7709481    64174757_15836021608_382578   start           32
6110525    756030933_38926415334_603627  start           32
1952790    961878360_85895454962_607105  start           32
2274658    518285126_8572733997_709761   start           32
5380759    694918984_92465751580_342876  start           32
8224170    273092462_14025162492_252779  start           32
7733877    308874104_3998249411_902973   start           32
7011180    750305338_48716927227_197559  start           32
6563017    943834869_97656130275_821170  start           26
5433263    393036510_74976230958_716895  start           25
3545814    416293838_23892754049_547527  start           24
5574879    875138661_34710212496_881092  step_2          24
                                         step_

In [23]:
# inspect time differences 

# convert time

df2["date_time"] = pd.to_datetime(df2["date_time"])

In [24]:
sample = duplicates_steps.sort_values(
    ["client_id", "visit_id", "process_step", "date_time"]
)

sample.head(10)

,client_id,visitor_id,visit_id,process_step,date_time
285315,336,64757908_3400128256,649044751_80905125055_554468,start,2017-06-01 07:26:55
285190,336,64757908_3400128256,649044751_80905125055_554468,start,2017-06-01 07:42:43
100903,786,439747392_30293505155,640294664_79662229470_499394,start,2017-06-01 11:00:22
100902,786,439747392_30293505155,640294664_79662229470_499394,start,2017-06-01 11:01:20
251290,805,831412807_82548325803,451173196_11661340552_563345,start,2017-06-15 19:09:13
251286,805,831412807_82548325803,451173196_11661340552_563345,start,2017-06-15 19:11:28
251288,805,831412807_82548325803,451173196_11661340552_563345,step_2,2017-06-15 19:09:26
251287,805,831412807_82548325803,451173196_11661340552_563345,step_2,2017-06-15 19:11:21
161966,832,645971208_10952400740,704999523_65579948030_431837,step_1,2017-06-16 10:38:47
161964,832,645971208_10952400740,704999523_65579948030_431837,step_1,2017-06-16 10:39:30


Customer-level analysis

In [29]:
# Unique customers

df2["client_id"].nunique()

# Number of visits per customer
visits_per_client = df2.groupby("client_id")["visit_id"].nunique()

visits_per_client.describe()

count    67430.000000
mean         1.237847
std          0.585019
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         17.000000
Name: visit_id, dtype: float64

Visit (Session)-level analysis

In [30]:
# Verify visit -> visitor relationship
df2.groupby("visit_id")["visitor_id"].nunique().value_counts()

# Verify visit -> client relationship
df2.groupby("visit_id")["client_id"].nunique().value_counts()

client_id
1    82219
2      617
3        5
Name: count, dtype: int64

Event-level analysis

In [31]:
# Repeated process steps
(
df2.groupby(
    ["client_id", "visit_id", "process_step"]
)
    .size()
    .sort_values(ascending=False)
    .head(10)
)

client_id  visit_id                      process_step
913770     512831152_67980938484_874337  start           48
5574879    875138661_34710212496_881092  step_1          40
8312319    356549486_83727228650_639957  start           33
7709481    64174757_15836021608_382578   start           32
6110525    756030933_38926415334_603627  start           32
1952790    961878360_85895454962_607105  start           32
2274658    518285126_8572733997_709761   start           32
5380759    694918984_92465751580_342876  start           32
8224170    273092462_14025162492_252779  start           32
7733877    308874104_3998249411_902973   start           32
dtype: int64

In [32]:
# check for duplicate

df2.duplicated().sum()

np.int64(8669)

In [8]:
# Basic descriptive stats
df2.describe(include='all')

,client_id,visitor_id,visit_id,process_step,date_time
count,4.122640e+05,412264,412264,412264,412264
unique,NaN,71042,82841,5,346697
top,NaN,722943003_3441581446,875138661_34710212496_881092,start,2017-05-02 10:53:01
freq,NaN,104,104,135035,24
mean,5.028227e+06,NaN,NaN,NaN,NaN
std,2.881828e+06,NaN,NaN,NaN,NaN
min,3.360000e+02,NaN,NaN,NaN,NaN
25%,2.533952e+06,NaN,NaN,NaN,NaN
50%,5.056314e+06,NaN,NaN,NaN,NaN
75%,7.516792e+06,NaN,NaN,NaN,NaN
